crates/runtime/src/memory.rs

```diff
@@ -254,35 +260,136 @@ impl RuntimeLinearMemory for MmapMemory {
+        size: usize,
+        make_accessible: Option<fn(*mut u8, usize) -> Result<()>>,
         memory_image: Option<MemoryImageSlot>,
-    },
+    ) -> Result<Self> {
+        if base.len() < size {
+            bail!(
+                "initial memory size of {} exceeds the pooling allocator's \
+                 configured maximum memory size of {} bytes",
+                size,
+                base.len(),
+            );
+        }
 
-    /// A "dynamic" memory whose data is managed at runtime and lifetime is tied
-    /// to this instance.
-    Dynamic(Box<dyn RuntimeLinearMemory>),
+        if let Some(make_accessible) = make_accessible {
+            if size > 0 {
+                make_accessible(base.as_mut_ptr(), size)?;
+            }
+        }
+
+        Ok(Self {
+            base,
+            size,
+            make_accessible,
+            memory_image,
+        })
+    }
+}
+
+impl Default for ExternalMemory {
+    fn default() -> Self {
+        Self {
+            base: &mut [],
+            size: 0,
+            make_accessible: Some(|_, _| unreachable!()),
+            memory_image: None,
+        }
+    }
 }
 
+impl RuntimeLinearMemory for ExternalMemory {
+    fn byte_size(&self) -> usize {
+        self.size
+    }
+
+    fn maximum_byte_size(&self) -> Option<usize> {
+        Some(self.base.len())
+    }
+
+    fn grow_to(&mut self, new_byte_size: usize) -> Result<()> {
+        // Never exceed the static memory size.
+        if new_byte_size > self.base.len() {
+            bail!("static memory size exceeded")
+        }
+
+        // Actually grow the memory.
+        if let Some(image) = &mut self.memory_image {
+            image.set_heap_limit(new_byte_size)?;
+            self.size = new_byte_size;
+        } else {
+            let make_accessible = self
+                .make_accessible
+                .expect("make_accessible must be Some if this is not a CoW memory");
+
+            // Operating system can fail to make memory accessible.
+            let old_byte_size = self.byte_size();
+            make_accessible(
+                unsafe { self.base.as_mut_ptr().add(old_byte_size) },
+                new_byte_size - old_byte_size,
+            )?;
+        }
+
+        // Update our accounting of the available size.
+        self.size = new_byte_size;
+        Ok(())
+    }
+
+    fn vmmemory(&mut self) -> VMMemoryDefinition {
+        VMMemoryDefinition {
+            base: self.base.as_mut_ptr().cast(),
+            current_length: self.size,
+        }
+    }
+
+    fn needs_init(&self) -> bool {
+        if let Some(slot) = &self.memory_image {
+            !slot.has_image()
+        } else {
+            true
+        }
+    }
+
+    #[cfg(feature = "pooling-allocator")]
+    fn into_any(self: Box<Self>) -> Box<dyn std::any::Any> {
+        self
+    }
+}
+
+/// Representation of a runtime wasm linear memory.
+pub struct Memory(pub(crate) Box<dyn RuntimeLinearMemory>);
```

Comment:
If possible, like above, I think it'd be best for this to not be `pub(crate)` but that may not be possible (if it's needed it's needed, can fix later)